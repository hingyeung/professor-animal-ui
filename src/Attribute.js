import React, { Component } from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.name = props.name;
  }

  render() {
    return (
      <div class="attribute-row">
        <div class="attribute-cell">
          {this.name}
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-yes'}>YES</label>
          <input type="radio" name={this.name} value="yes" id={this.name + '-yes'}/>
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-no'}>NO</label>
          <input type="radio" name={this.name} value="no" id={this.name + '-no'}/>
        </div>
      </div>
    )
  }
}

export default Attribute;