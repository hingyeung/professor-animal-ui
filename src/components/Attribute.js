import React, {Component} from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.props = props;
    this.attribute = props.attribute;
    this.type = props.type;
    this.onAttributeChange = props.onAttributeChange;
  }

  onChange(attrType, attrName, e) {
    this.onAttributeChange(attrType, attrName, e.target.value);
  }

  render() {
    const yesId = this.type + '-' + this.attribute.name + '-yes',
          noId = this.type + '-' + this.attribute.name + '-no';

    return (
      <tr className="attribute-row row">
        <td className="attribute-cell col-sm-6">
          { this.attribute.name }
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="yes"
                 id={ yesId }
                 onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={this.attribute.value}/>
          <label className="form-check-label" htmlFor={ yesId }>YES</label>
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="no"
                 id={ noId } onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={!this.attribute.value}/>
          <label className="form-check-label" htmlFor={ noId }>NO</label>
        </td>
      </tr>
    )
  }
}

export default Attribute;