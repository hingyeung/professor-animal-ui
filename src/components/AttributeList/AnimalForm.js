import React, {Component} from 'react';
import AttributeGroup from "./AttributeGroup";

class AnimalForm extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    let attributeGroupContent = [];

    const currentAnimal = this.props.animal;
    Object.keys(currentAnimal.attributeMap).forEach(attributeType => {
      attributeGroupContent.push(
        <AttributeGroup key={ attributeType + currentAnimal.id }
                        attributeType={ attributeType }
                        attributes={ currentAnimal.attributeMap[attributeType] }/>
      );
    });

    return (
      <form onSubmit={ this.props.onFormSubmit }>
        <div>
          <button className="btn btn-primary">Save</button>
        </div>
        <div className="form-group row">
          <label htmlFor="animal-name" className="col-sm-2 col-form-label">Animal name</label>
          <div className="col-sm-6">
            <input id="animal-name" className="form-control" type="text" name="animalName"
                   onChange={ (e) => this.props.updateAnimalName(e.target.value) } value={ currentAnimal.name }/>
          </div>
        </div>
        { attributeGroupContent }
        <div>
          <button className="btn btn-primary">Save</button>
        </div>
      </form>
    )
  }
}

export default AnimalForm;