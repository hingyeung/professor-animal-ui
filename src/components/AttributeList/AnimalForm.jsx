import React from 'react';
import AttributeGroup from "./AttributeGroup";

function AnimalForm(props) {
  const onAttributeChange = props.onAttributeChange,
    onFormSubmit = props.onFormSubmit,
    updateAnimalName = props.updateAnimalName,
    animal = props.animal;

  let attributeGroupContent = [];

  const currentAnimal = animal;
  Object.keys(currentAnimal.attributeMap).forEach(attributeType => {
    attributeGroupContent.push(
      <AttributeGroup key={ attributeType + currentAnimal.id }
                      animalId={ currentAnimal.id }
                      onAttributeChange={ onAttributeChange }
                      attributeType={ attributeType }
                      attributes={ currentAnimal.attributeMap[attributeType] }/>
    );
  });

  return (
    <form onSubmit={ onFormSubmit }>
      <div>
        <button className="btn btn-primary">Save</button>
      </div>
      <div className="form-group row">
        <label htmlFor="animal-name" className="col-sm-2 col-form-label">Animal name</label>
        <div className="col-sm-6">
          <input id="animal-name" className="form-control" type="text" name="animalName"
                 onChange={ (e) => updateAnimalName(e.target.value) } value={ currentAnimal.name }/>
        </div>
      </div>
      { attributeGroupContent }
      <div>
        <button className="btn btn-primary">Save</button>
      </div>
    </form>
  )
}

export default AnimalForm;