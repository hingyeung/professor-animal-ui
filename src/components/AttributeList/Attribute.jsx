import React from 'react';

function Attribute(props) {
  const attribute = props.attribute;
  const type = props.type;
  const onAttributeChange = props.onAttributeChange;
  const yesId = props.type + '-' + props.attribute.name + '-yes',
    noId = props.type + '-' + props.attribute.name + '-no',
    // Initially this.state.attributeValue is undefined (before animals are loaded),
    // and it makes the radio button "uncontrolled", then later when this.state.attributeValue
    // is populated with either true or false and the radio button becomes "controlled".
    // ReactJS doesn't like it and would give warning. Making sure its value is either true
    // or false prevents the problem.
    // https://github.com/facebook/react/issues/6779#issuecomment-222162404
    yesIsChecked = props.attribute.value ? props.attribute.value : false;

  function onChange(e) {
    onAttributeChange(type, attribute.name, e.target.value);
  }

  return (
    <tr className="attribute-row row">
      <td className="attribute-cell col-sm-6">
        { attribute.name }
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="yes"
               id={ yesId }
               onChange={ onChange }
               checked={ yesIsChecked }/>
        <label className="form-check-label" htmlFor={ yesId }>YES</label>
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="no"
               id={ noId } onChange={ onChange }
               checked={ !yesIsChecked }/>
        <label className="form-check-label" htmlFor={ noId }>NO</label>
      </td>
    </tr>
  )
}

export default Attribute;