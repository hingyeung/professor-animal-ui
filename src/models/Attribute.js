class Attribute {
  constructor(name, defaultValue) {
    this._name = name;
    this._defaultValue = defaultValue;
    this._value = '';
  }

  get name() {
    return this._name;
  }

  get defaultValue() {
    return this._defaultValue;
  };

  get value() {
    return this._value;
  }

  set value(value) {
    this._value = value;
  }
}

export default Attribute;