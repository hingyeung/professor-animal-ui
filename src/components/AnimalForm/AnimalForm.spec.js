import React from 'react';
import {shallow, mount} from 'enzyme';
import AnimalForm from "./AnimalForm";
import Attribute from 'models/Attribute';

describe('AnimalForm', function () {
  const animalId = "animalId";
  const animal = {
    id: animalId,
    name: "animal1",
    attributeMap: {
      physical: {"attributeA": true},
      diet: {"attributeB": true}
    }
  };
  const mockOnFormSubmit = jest.fn();

  beforeEach(function () {
    mockOnFormSubmit.mockClear();
  });

  it('should render properly', function () {
    const wrapper = shallow(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);
    expect(wrapper).toMatchSnapshot();
  });

  it(`should not call onFormSubmit when animal name is not filled and the Save button is clicked`, function () {
    const wrapper = shallow(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);
    // The code under test relies on onChange event, simply filling in the text field using Enzyme doesn't
    // trigger the event.
    // https://github.com/airbnb/enzyme/issues/76#issuecomment-189606849
    wrapper.find('input#animal-name').simulate('change', {target: {value: ''}});
    wrapper.find('form').simulate('submit', {preventDefault: jest.fn()});

    expect(mockOnFormSubmit).toHaveBeenCalledTimes(0);
  });

  it('should call onFormSubmit when animal name is filled and the form is submitted', function () {
    const wrapper = shallow(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);

    wrapper.find('input#animal-name').simulate('change', {target: {value: 'animal name'}});
    wrapper.find('form').simulate('submit', {preventDefault: jest.fn()});

    expect(mockOnFormSubmit).toHaveBeenCalledTimes(1);
    let expectedAnimal = Object.assign({}, animal);
    expectedAnimal.name = 'animal name';
    expect(mockOnFormSubmit).toHaveBeenCalledWith(expectedAnimal);
  });

  it('should update animal in state on attribute value change', function () {
    const wrapper = mount(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);

    wrapper.instance().onAttributeChange('physical', 'attributeA', false);
    expect(wrapper.state().animal.attributeMap.physical.attributeA).toBe(false);
  });

  it('should update animal in state on new attribute added', function () {
    const wrapper = mount(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);
    wrapper.instance().onNewAttributeAdded('physical', new Attribute('attributeC', true));
    expect(wrapper.state().animal.attributeMap.physical.attributeC).toBe(true);
  });

  it('should set animal in state when new animal received from props and the new animal has a different ID', function () {
    const wrapper = shallow(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);
    const newAnimal = {name: 'new animal', id: 'new id', attributeMap: {}};
    wrapper.setProps({animal: newAnimal});
    expect(wrapper.instance().state.animal).toEqual(newAnimal);
  });

  it('should not set animal in state when new animal received from props and the new animal has the same ID', function () {
    const wrapper = shallow(<AnimalForm animal={ animal } onFormSubmit={ mockOnFormSubmit }/>);
    const newAnimal = {name: 'new animal', id: animalId, attributeMap: {}};
    wrapper.setProps({animal: newAnimal});
    expect(wrapper.instance().state.animal).toEqual(animal);
  });
});