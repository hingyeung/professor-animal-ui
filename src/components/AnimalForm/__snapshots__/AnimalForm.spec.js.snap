// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalForm should render properly 1`] = `
<div
  className="animal-form-container container"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group row"
    >
      <label
        className="col-sm-2 col-form-label"
        htmlFor="animal-name"
      >
        Animal name
      </label>
      <div
        className="col-sm-6"
      >
        <input
          className="form-control"
          id="animal-name"
          name="animalName"
          onChange={[Function]}
          type="text"
          value="animal1"
        />
      </div>
      <div
        className="col-sm-4"
      >
        <button
          className="btn btn-primary ml-5"
          type="submit"
        >
          Save
        </button>
      </div>
    </div>
    <AttributeGroup
      attributeType="physical"
      attributes={
        Object {
          "attributeA": true,
        }
      }
      key="physicalundefined"
      onAttributeChange={[Function]}
      onNewAttributeAdded={[Function]}
    />
    <AttributeGroup
      attributeType="diet"
      attributes={
        Object {
          "attributeB": true,
        }
      }
      key="dietundefined"
      onAttributeChange={[Function]}
      onNewAttributeAdded={[Function]}
    />
    <div>
      <button
        className="btn btn-primary"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
